### ISSUE TITLE
Implementation by delegation enhancements

### ISSUE THREAD

**Issue by Dico200** [Kotlin/KEEP#155]
This is an issue to discuss proposal for [Implementation by delegation enhancements](https://github.com/Dico200/KEEP/blob/ibd-enhancements/proposals/implementation-by-delegation-enhancements.md) (PR #154)

**Comment by voddan** [comment]
> Add a compiler-intrinsic way to access the delegate used for a given interface:
> inline fun <reified T> Any.delegate(): T

This particular signature won't work because it needs to specify which delegate you want to access.  
A class can  implement any number of interfaces implemented by delegates:

    interface A
    interface B
    interface C

    val a = object : A {}
    val b = object : B {}
    val c = object : C {}

    class X() : A by a, B by b, C by c

**Comment by voddan** [comment]
## Syntactical proposal
- Deprecate `X() : A by a` and introduce a *new keyword* for class delegation with improved behaviour
- Use `to` as a contextual keyword for class delegation: `X(): A to a` 

### Pros
- Syntactically backward compatible: since no expression was allowed in the position after the type a class is implementing `X(): A >here<`, anything can be a soft keyword without clashing with old code 
- `to` is short: just 2 characters
- Has some meaning in English: "`X` delegates `A` *to* `a`"
- Syntax differs from the property delegation syntax and stops confusing newcomers with a concept of delegation having two totally different use cases

### Cons
- If implemented would allow two syntax constructs to express the same thing (even if one is deprecated):
```
class X(val a: A): A by a
class X(val a: A): A to a
```
- Syntax differs from the property delegation syntax and doesn't highlight that those are similar concepts

**Comment by Dico200** [comment]
@fvasco If you're alluding to the error that would be generated before the `lateinit` property is initialized, that would be an error on the part of the programmer which you can get with any implenentation of the delegation pattern. When the programmer uses `lateinit`, they must understand the implications. We should not prevent the programmer from writing buggy code by heavily restricting them. If we did, `lateinit` would arguably not exist.

**Comment by fvasco** [comment]
No,
it is another proposal, which I prefer.
I perfectly agree your last statement.

I consider it more flexible and it solves some other issues.

Are you considered it?

**Comment by Dico200** [comment]
Well, not specifically that syntax, but I do class it among the other options for having a separate syntax. And my opinion on it is thus the same. Feel free to convince me that it is better than having a combination of one annotation and a compiler argument to phase out the old behaviour.

**Comment by voddan** [comment]
> the delegate function was proposed to have a reified type parameter to specify the delegate's interface

You are right I didn't pay enough attention to the signature. Sorry.

>  please see the proposal's Implementation section for my reasoning NOT to add a new syntax. I think that's a very bad idea.

In the proposal you just state that you "do not recommend the different syntax approach", which is a personal opinion, not an objective reason. I personally think that the syntax I proposed is the best we can hope for, and I gave pros and cons of this syntax. Please add those to the proposal as one of the possible options. Similarly, please add *detailed* pros and cons for other syntactic options.

> Sorry but I get the impression you haven't actually read the proposal in full

Actually I had read the proposal from top to bottom before commenting, and still had a lot of questions. 

As a side note, this particular proposal is very hard to read because it has  a lot of "water" in it, e.i. phrases that are there for stile and not for content. A KEEP proposal is a technical document, not an essay, so it should be as concise and as straight forward as possible. I recommend watering it down so that people have easier time reading it. You may use JetBrains-authored proposals as examples.

**Comment by voddan** [comment]
@Dico200 wrote to @fvasco: 
>  Feel free to convince me that it is better than having a combination of one annotation and a compiler argument

IMHO it is not our goal to convince you in anything. As far as I understand it, you are the author of the KEEP, so your role is to collect **all** possible options for solving this particular problem, analyse them and write down all pros and cons and trade-offs and reasons. Then you and @fvasco and me and others go to the core Kotlin designers together and try to convince them that this proposal makes sense. 

That's how I see this process, please correct me if I am wrong.

**Comment by Dico200** [comment]
@voddan Alright, that makes sense. You are correct. I've added your proposed syntax to the list of syntaxes that have been proposed. Sorry for the bad attitude.
I'm rewording the *Implementation* section to remove the clear signs of bias that I added and allow the reader to see the arguments in favor and against both of the approaches.

As for watering it down - I don't know how I would do that without removing important notes and/or information. Can you give an example of how you would rewrite a section?

Finally, the reasons against having a new syntax follows directly after the part you quoted.

**Comment by Dico200** [comment]
I've also improved the language and ambiguity of the document a little bit, and tweaked the policy described in *JVM Codegen*

**Comment by Dico200** [comment]
@fvasco I realize that I misunderstood your last comment.
I don't understand your proposal with regards to `lateinit`. As far as I'm concerned, both options you show are valid with the new behaviour, so you would be free to do that.

**Comment by Dico200** [comment]
@voddan I've quoted your arguments in favor and against `A to b` syntax in the Approach section.

**Comment by voddan** [comment]
@Dico200 Thanks! I'll try to "reword" some of the proposal later today and send you a PR

**Comment by Dico200** [comment]
I have clarified and refined the discussion of proposal 2 in the file.
It becomes very clear that there are really just 2 almost completely separate proposals intertwined in the file, which raises the question if there should be 2 separate KEEPs. But I think it is fine for now.

**Comment by fvasco** [comment]
I quick answer myself and I put some minor consideration on the table

```kotlin
interface A
class A1 : A

val a = A1()

class X(val a: A) : A by a
```

This code is already valid in Kotlin, so the proposals using `by a`, `by val a` and `by volatile a` do not work as expected. All of these reference to an `a` instance outside the class `X`.

Moreover the below syntax is alread valid

```kotlin
class X2(val a: A) : KProperty<A1> by ::a
```

so the proposal `by ::a` is an incompatible change.

**Comment by voddan** [comment]
> t becomes very clear that there are really just 2 almost completely separate proposals intertwined in the file, which raises the question if there should be 2 separate KEEPs.

True. I think we should split the KEEPs so that if one is rejected, the other could still be accepted.

**Comment by Dico200** [comment]
@fvasco In the current syntax, the below is also valid:

```kotlin
interface A
class A1 : A

val a = A1()

// note that the parameter is not assigned to a property
class X(a: A) : A by a
```

This is because the only things visible from the delegate expression are the constructor parameters, (and file scope, global scope).

With the new behaviour, this is invalid, because constructor parameters are *not* visible. Instead, `this` instance is visible. With conflicting declarations between class scope and outer scope, class scope is prioritized, as it would be from any expression within the class.

With that out of the way, the programmer must make a conscious decision to use the new behaviour according to the proposal. This is why it's okay for the behaviour to be different than they expect - they need to explicitly opt in to have the behaviour change, whether by using a different syntax, adding an annotation or using a compiler argument.

So when the programmer says `by val a` instead of `by a`, they are choosing the behaviour explicitly.
It's a bit like saying `listOf` vs `setOf` when you want a `Collection`, although that analogy isn't the most accurate, I must say.

As for the infeasibility of `::a` syntax, that's correct. I do believe the file mentions that.

**Comment by voddan** [comment]
>  they need to explicitly opt in to have the behaviour change [by] using a compiler argument

@Dico200 a compiler argument can't be a part of the solution because it would conflict with keeping backward compatibility. 

To illustrate, imagine an old code base which uses old syntax. Source backward compatibility in essence means that you may update your compiler (nothing breaks) and start using new features (nothing breaks). The moment you have a compiler argument that changes old behavior to new behavior, you can't use it in a legacy code base without potentially breaking it, and thus loose backward compatibility.

**Comment by Dico200** [comment]
Yeah, good point. I touched on that in the *Deprecation* section but I guess it applies here too.

**Comment by Dico200** [comment]
I added another approach for proposal 1, which is a lot like
`implement A by b` in that the delegation is declared inside the class body,
but it leverages kotlin properties and their semantics much better, making it very intuitive to use.

Thus, I am considering dropping the alternatives or just putting them at the bottom file in a section where we discuss the other options that we considered.

Before I do that though, I would like to have some feedback. Please look at the considerations in the file.
@voddan

**Comment by tieskedh** [comment]
@Dico200 

I love it!

```kotlin
interface A
class B(map : Map<String, Any>) : A, B, C{
    delegate val a : A by map
    delegate val b : B by map
    delegate val c : C by map
}
```

**Comment by Dico200** [comment]
@tieskedh Let's dissect that a little bit, and see if I understand what you're showing us.

1. To be clear, the proposal does not require `by` keyword in `delegate val`.
2. Thus, you're delegating an interface to a delegated property.
3. Kotlin `Map<in String, *>` is `ReadOnlyProperty`, whose `getValue` returns the map's `map["property name"]
4. So the interfaces are effectively delegated as such:
```
delegate val a : A get() = map["a"] as A
delegate val b : B get() = map["b"] as B
delegate val c : C get() = map["c"] as C
```

Is this correct? If it is, I'm glad you like it!

**Comment by voddan** [comment]
The last idea defenetly looks very interesting.

I would not recommend dropping alternative options thou, since we haven't had any feedback from JB guys, and they may have their reasons for tabuing some of the approaches.

**Comment by Dico200** [comment]
Fair enough @voddan, I agree.
I added (WIP) marker earlier so we can have a chance to discuss that before JB guys look at it. I will drop it now.

**Comment by tieskedh** [comment]
@Dico200, I know, and I like it!

**Comment by fvasco** [comment]
Reusing a kotlin property syntax for delegation may be a nice idea, but I do not throw declaration anywhere in the class body.

Class inherit behavior from delegated instance like class inheritance, so I consider usefull describe the delegation in the class definition, like the current behaviour, using the same solution of the `where` clause.

```kotlin
class D(val map : Map<String, Any>) : A, B, C
    delegate val a : A get() = map["a"] as A,
             val b : B get() = map["b"] as B,
             val c : C get() = map["c"] as C {
```

**Comment by tieskedh** [comment]
I liked the former proposal more, because it looks like they are normal properties, with delegation as an extra feature. Therefor, I think this will look slightly more confusing.
```kotlin
class D(val map : Map<String, Any>) : A, B, C delegate 
    open var a : A 
        get() = map["a"] as A,
        set(value) {
            if(value.isFoo()) field = createA() 
        } 
    protected abstract val b : B get() = map["b"] as B,
     val c : C get() = map["c"] as C {
```

I can understand however why you want it to be part of the header, because I always wanted the same thing. 

Ps. are you allowing to use this in the context? 
If not, you create boilerplate.
If yes, why is it allowed? You're not in the body of the class. Also why would it be allowed here and not using `by`?

**Comment by fvasco** [comment]
Hi @tieskedh ,
I agree with you.

Maybe we should consider a better idea.

**Comment by tieskedh** [comment]
I think the idea is pretty good.
We only need to keep the delegated properties as part of the implementation and not of the inheritance.
Another solution would be to add a syntax like companion object for delegates, but I'm sure it would be changed to the current proposal after a while, as companion funs are requested at this moment as well. (although I don't like it, but that's another discussion...)

**Comment by Dico200** [comment]
@fvasco I dont think delegation should be a part of the class header at all.
When you implement delegation manually, you do it by writing an implementation of all the public members that delegates explicitly to a member of the class. You put all of that, obviously, inside the class body.

On top of that, delegation has no implication whatsoever with regards to the type of the class and the interfaces it implements. It is a detail of how an interface is implemented - An *implementation detail*. Implementation details should be encapsulated. Moreover, I think it's obnoxious to have to put lots of code before the opening `{` bracket of the class body.

**Comment by Dico200** [comment]
@tieskedh Be careful when you say `delegated properties` as that's a term reserved for `by` syntax for properties.
Property delegates are delegating the implementation of a property to an object (by delegating `getValue` and `setValue` only).
They are different in the sense that they get extra information, such as who is delegating and which property is actually delegated.

You can achieve almost exactly the same with interface delegates. Just delegate an interface that declares a property, and delegate the interface.
The key difference, again, is that property delegates get a bit of extra information, useful information. They are actually aware of their function as a delegate.  
This is because, to mimic a property delegate, you call `delegate.getValue(this, MyClass::thisPropertyName)` from the getter.  
In the case of interface delegates, you would return `delegate.thisPropertyName`, which, quite clearly, doesn't give that information.

Another difference is that, when using an interface, that interface decides the declaration (name, type), whereas property delegates don't do that.

In the end, property delegates and interface delegates have very different use cases.
It's complex, they are kinda similar, but I don't think they should use the same keyword because of it. 

So, what term should you use when referring to `delegate val`? I would say "Delegate property" works well.

**Comment by tieskedh** [comment]
haha, again a mistake that makes me look like a starter.
Maybe Implementation-property?

**Comment by Dico200** [comment]
It's not really a mistake, just an easy confusion.
Some other keywords: `interface val`, `implementation val`, `implement val`, `decorator val`, `delegating val` 

Another syntax:
```kotlin
val listDelegate: ArrayList<Int>
    delegate List<Int>
    get() = emptyList()
```

**Comment by tieskedh** [comment]
1st suggestion, I don't like it, as I think delegate is a very important part. now it's just alined with getters and setters.

2nd suggestion in former code doesn't allow access to this (without confusing)

**Comment by voddan** [comment]
>1st suggestion, I don't like it, as I think delegate is a very important part. now it's just alined with getters and setters.
>
>2nd suggestion in former code doesn't allow access to this (without confusing)

@tieskedh Sorry, could you copy or explain the statements you are agreeing or disagreeing with? It is a long discussion and it is hard to keep track of what is 1st or 2nd suggestion.

**Comment by tieskedh** [comment]
updated. 
I agree.
My bad

**Comment by fvasco** [comment]
Hi @Dico200

> You put all of that, obviously, inside the class body.

This is the only way to do that now: delegation in the header and implementation in the body.

We are talking about new language features, and new syntax should be motivated.

I consider delegation and inheritance very similar (sometimes interchangable).
I vote for mantaining the current Kotlin behaviour: delegation in declaration site.

I constantly considering a syntax like:

```kotlin
class C : A by &a {

    val a : A ...

}
```

but I am looking for a better choice.
(Proposal "Using Different Syntax" is backward incompatible.)

**Comment by Dico200** [comment]
@fvasco 
I don't agree that delegation and inheritance are similar.
Using a different syntax is not backward incompatible at all, as long as it's distinguishable from the old syntax.
We'll see what JB thinks.

**Comment by tieskedh** [comment]
@fvasco
 I started to like the previous syntax a lot, but if I have to choose, I choose your syntax. 
Here is why:

- It's the same concept
- it uses familiar syntax
- you can use `this`
- you can use the delegating without having a class-body
- there is no need to deprecate the old syntax
- the functionality from the delegate-val can still be used, but just doesn't have a marker. 
    We can however require a special marker on the property.

In short, I find your last syntax the best until now

**Comment by Dico200** [comment]
Shall we just make a channel in the kotlin slack for this discussion?

**Comment by fvasco** [comment]
You cannot use `this`, it is not retro compatible.

```kotlin
interface A
interface B
interface C

class AImpl : A, C
class BImpl : B, C

object O {
    val a = AImpl()
    val b = BImpl()

    class X() : A by this.a, B by this.b, C by this.a // this is the O instance
}
```

The above code already works in Kotlin

**Comment by voddan** [comment]
@fvasco Sorry, wat? Do you mean my pseado-code which I used to illustrate the potential issue with the new syntax proposed in *second approach* ?

**Comment by fvasco** [comment]
Hi @voddan,
in the code `class X() : A by this.a` the `this` reference does not refer to the `X` instance, in my example is the `O` instance, so we cannot use `this` outside the class body: it is a breaking change.

Finnally the `a` property in your example is of type `A`, and the interface`A` does not extend the interface `C`.

That's all.

**Comment by tieskedh** [comment]
@voddan 
You're right.
Seems like a lot of boilerplate when you add delegate in front.

**Comment by voddan** [comment]
@tieskedh 
Not just boilerplate. I can't figure out how to do it at all. My best attept with the proposed syntax:

    class X() : A, B, C
    {
        delegate val a: A = AImpl() // implements A
        delegate val b: B = BImpl() // implements B
        delegate val c: C = a       // implements C??
    }

**Comment by tieskedh** [comment]
@voddan

It would be boilerplate:

class X() : A, B, C
{
    delegate val a: A = AImpl() // implements A
    delegate val b: B = BImpl() // implements B
    delegate val c: C 
         get() = a
         set(value) {
             a = value
         }    // implements C
}

**Comment by voddan** [comment]
@tieskedh so the proposed rule is that we look at the type of the property `val c: C` to figure out which interface it implements, correct? I think it should be written explicitly in the corresponding section of the proposal.

**Comment by tieskedh** [comment]
My mistake, updated

**Comment by voddan** [comment]
@Dico200 I think we need to delete the second part of the proposal "Add a compiler-intrinsic way to access the delegate identity " and instead add it as one of the requirements we wont from an implementation. Then different approaches will solve the problem of accessing the delegate in their own ways (via an intrinsic function or just as a property accessor). That will simplify the text without affecting the end result.

**Comment by Dico200** [comment]
@voddan 
> I think we need to delete the second part of the proposal "Add a compiler-intrinsic way to access the delegate identity " and instead add it as one of the requirements we want from an implementation. 

That makes sense, I'll look into a structured way to do that. Might move the second proposal into a separate gist and link to it. What do you think?

**Comment by Dico200** [comment]
@tieskedh 
> ```kotlin
> class X() : A, B, C
> {
>     val abImpl = ABimpl() 
>     delegate val a: A by this::abImpl// implements A
>     delegate val b: B = BImpl() // implements B
>     delegate val c: C by this::a BImpl //implements 
> }
> ```

Property delegation is illegal here, you can't delegate a property to a `KProperty`.
But you got the idea right - just need to return the value from another property to delegate 2 interfaces to the same property. See my comment above: just use a getter.

**Comment by Dico200** [comment]
@fvasco 
> You cannot use `this`, it is not retro compatible.

Very good point, you are correct, in the context of using the same syntax as old behaviour.  
However, it is perfectly fine if `by val` or `to` is introduced as keywords for the new behaviour.  
Compiler and programmer can always distinguish `by val` from the old `by` syntax.  
@voddan Did use old `by` syntax in his examples, but I think he used it as a placeholder for any of the choices put forward in *Approach 1* for proposal 1.

**Comment by Wasabi375** [comment]
What about this kind of syntax
```kotlin
class X: A, B, C {
    delegate A val a = someImpl // implements A
    delegate B, C val b = otherImpl // implements B and C
}
```
I don't like the idea of assigning the delegates just based on the type of the property. Also this is a shorter way to state that one object is used for multiple interfaces.

**Comment by Dico200** [comment]
@Wasabi375
> What about this kind of syntax

When discussing the option with my friend, we did consider that, but we thought just `delegate val` is better because it is simpler. Just a modifier of a property, really. I would describe it as "more kotliny".
That aside, I do think that idea is viable. On the other hand, I think having one `delegate` property for every delegated interface makes sense and is also perfectly acceptable. And when that rule exists, using the type of the property to indicate which interface is to be delegated makes sense, for simplicity of code.

**Comment by Dico200** [comment]
When typing what I just said, I realized there's a possible readability issue when the property's type is omitted, because it is inferred. In that case, it might become very unreadable. Maybe a `delegate val` must always be required to declare type explicitly (if we use that approach).

**Comment by fvasco** [comment]
I rewrite @Wasabi375 proposal using `get()`/`set()` like syntax.

```kotlin
class X: A, B, C {
    val a = someImpl // implements A
      delegate A

    val b = otherImpl // implements B and C
      delegate B, C 
}
```

**Comment by Dico200** [comment]
I already put that forward as well @fvasco 
> ```kotlin
> val listDelegate: ArrayList<Int>
>     delegate List<Int>
>     get() = emptyList()
> ```

Though I just realized that `emptyList()` is not an `ArrayList`, oops.

**Comment by Dico200** [comment]
It's clear that we need to expand the proposal further to accommodate the suggested ideas, and to make approaches individually responsible for explaining how a delegate can be accessed from the code. *Proposal 2* is only relevant for the old behaviour and approaches *Approach 1* and *Approach 3* for *Proposal 1*, confusing the file.

**Comment by Dico200** [comment]
Type restriction of `delegate val` could be made more lenient, any type as long as there is exactly one common interface between it and the declaring class.

**Comment by tieskedh** [comment]
@Dico200 
I don't like that idea at all.
There are a lot of ideas, but they have all one thing in common:
It's immediately clear which variable delegates to wich interface.
I would definately disaprove a syntax, where this isn't the case.

**Comment by voddan** [comment]
>>I think we need to delete the second part of the proposal "Add a compiler-intrinsic way to access the delegate identity " and instead add it as one of the requirements we want from an implementation.
>
>That makes sense, I'll look into a structured way to do that. Might move the second proposal into a separate gist and link to it. What do you think?

@Dico200 If you mean that you want to form another KEEP for this additional proposal, I would advise against it since without the current proposal such a delegate assessor would have very limited usage.

As for gists for previewing the next iteration of the proposal, I usually prefer looking at diffs in commits or PRs. For example you can create a separate branch with the updated document, then create a PR to this branch and link the PR here. Or just do the editing and link here the commit :)

**Comment by voddan** [comment]
I carefully looked at the suggested syntax options which use the class body to declare delegation, and I think they have one common disadvantage: the delegating construct is very small and easily lost in the clatter of other class members (e.i. properties, methods, subclasses). 

In the syntax with property modifiers `delegate val a: A` the delegation property can be placed in any place in a class, potentially very far from the top in the middle. In other syntactic options where there is a separate delegation construct like `delegate A to a`, there is no clear rule where this construct should be put, so again it can end up in the bottom of the class declaration. All of those scenarios dramatically harm the readability of the code.

Because of that I start to incline to the @fvasco's option of putting the delegation construct into the `where` clause, which is **after** the main class signature, but **before** the main class body:

    class X : A, B, C by C(), D by newD()
        where A delegates a, B delegates this.b 
    {
        var a =  A()
        var b = B()
    }

This approach simultaneously: simplifies the main class signature + keeps the delegation constructs at the top + incorporates the old syntax as a short cut for some of the delegations + allows for a new semantics for the delegations that need recalculating their delegate. IMHO it's nearly perfect.

What do you think?

P.S. @Dico200 Please kindly put both the argument against moving declaration inside the class body and the described syntax option into the KEEP.

**Comment by fvasco** [comment]
Hi @voddan 
I really agree with your concerns, but I do not mix `where` and `delegate` parts.

Moreover I remark that it is not possible using `this` outside the class body, because it does not refer to the `X` instance.

If you refer to KT-83 we are using this KEEP to solve two issues: mutate the delegation reference (the `volatile proposal`) and refer to a class's property (the `val` proposal). Obviously we can solve both issue using an unified way.

I propose here a little example, I hope this can be useful to battle test further proposals.

```kotlin
interface A
interface B
interface C
interface D

class A1:A
class B1:B
class C1:C
class D1:D


object Scope {
    var a = A1()

    class X(b:B = B1()) :
            A by this.a, // delegation to Scope.a, non mutating reference
            B by b, // delegation to parameter
            C by val C, // delegation to local property (non mutating reference), we have to define the initialization time
            D by var D { // delegation to local property (mutating reference)
        var c=C1
        var d=D1
    }
}
```

The above example does not solve the problem: mutating reference for local variable delegation, however it is pretty easy to work around this issue

```kotlin
import kotlin.reflect.KProperty

interface A

class A1 : A

fun main(args: Array<String>) {
    var a = A1()
    val mutatingDelegation by { a }
}

inline operator fun <T> (() -> T).getValue(thisRef: Any?, property: KProperty<*>): T = invoke()

```

**Comment by voddan** [comment]
Hi @fvasco!
Sorry if I misinterpreted your suggestion. It looked to me more natural to re-use the existing `where` section than to create a totally new one.

>  it is not possible using `this` outside the class body

That is true now, but if we add new syntax, we are free to change this rule. I used `B delegates this.b` in my snippet to highlight that `this` is available and refers to the instance itself, and not some outside object. Similarly `A delegates a` in my mind should refer to any property of the class.

That said I am worried that inconsistent `this` rules will be, well, inconsistent. So maybe requiring explicitly pointing to the class  with a label `A delegates this@X.a` is a better option.

In the context of my proposed syntax your testing example would look something like the following:

    object Scope {
        var a = A1()

        class X(b: B = B1()) : A by this.a, B by b, C, D 
        where C delegates c, D delegates this@X.d // <--- same semantics. Mutability depends on the property itself
        {
            val c = C1
            var d = D1
        }
    }

**Comment by tieskedh** [comment]
a couple of **my** opinions:

- Putting `this` outside a class can confuse, so should be avoided
- You should be able to use  `this` for properties straight away
- The delegation should be easy to find
- The syntax should be related to the previous syntax, because it's the same concept.
   this means (part of) the delegation needs to happen in the class-header.
- The syntax should be easy to understand.

I want to focus again on this syntax, as I find it very elegant:

```kotlin
class C : A by &a {
    val a : A ...
}
```

or with the previous example:

```kotlin
object Scope {
    var a = A1()

    class X(b: B = B1()) : A by a, B by b, C by this@X.c, D by &this@X.d {
        val c = C1
        var d = D1
    }
}

```
Where & can be added or removed for every delegate, depending on the needs.
Maybe, just maybe, you could refer to a variable inside the class using `in`.

and to answer the question about delegation to propperties:

```kotlin
private inline operator fun <reified R> KProperty0<R>.getValue(any: Any?, property: KProperty<*>) = this.get()
private inline operator fun <reified R> KMutableProperty0<R>.setValue(any: Any?, property: KProperty<*>, value: R) = this.set(value)
```

are there any drawbacks uing this syntax?

**Comment by voddan** [comment]
@tieskedh 
>> class C : A by &a 
>
>are there any drawbacks uing this syntax?

Syntactically, using a single character is not a very good idea because the reader has to know what it means or they have no hope of googling it. You are probably coming from C/C++ where `&` is a very common symbol and means reference, which is somewhat relevant here. But for the rest of the world `&a` is a totally unfamiliar notation. For that reason Kotlin generally prefers using keywords instead of single characters where it can. For your notation it could look like `class C : A by volatile a` (from the KEEP).

Anyways this syntax falls in the "1. Using Different Syntax" group and has similar pros and cons as the other options in this group. I suggest you add your particular syntax to the KEEP together with arguments why you think it is better than the other options.

**Comment by tieskedh** [comment]
@voddan I know it from PHP, thanks for explaining.

**Comment by voddan** [comment]
> @voddan I know it from PHP, thanks for explaining.

Ow, I didn't know they had it in PHP :) 
My previous comment is just one argument against this particular syntax. IMO for a comprehensive discussion we should consider all possible options, so I would very much like seeing your proposed syntax as one of the options in the text of this KEEP proposal.

**Comment by tieskedh** [comment]
I personally feel like the KEEP is directed towards one syntax: approach 2.

In the beginning it says: the aim is to have the same or almost the same semantics as kotlin properties.

Then it says that we add a new language feature.
It can be seen like that, but I see it as an extension of the current one. This way of viewing is not really approved be the keep, as later on the by-word is disapproved twice:

pro: 
- In the case of [2], syntax doesn't borrow by keyword, so it doesn't _incorrectly_ highlight that those are similar concepts
con:
- We add a new language feature, instead of reusing old syntax

**Comment by Dico200** [comment]
I don't recall where it was said but:
I disagree that declaration of the delegate must be found at the top of the class definition, that it's important to see it easily. Why can't it be in the middle of the class? Again, it is an implementation detail.
This detail should be documented by the programmer if it's relevant, but maybe it's not relevant at all for  an api user that some object is delegating an interface.

**Comment by Dico200** [comment]
I've improved the structure of the proposal, it will be simpler to add new approaches without confusing everything.

**Comment by Dico200** [comment]
@voddan 
> What do you think?

Here's my honest opinion. I disagree with you on everything you said in this comment. Below are my arguments.

> I think they have one common disadvantage: the delegating construct is very small and easily lost in the clatter of other class members (e.i. properties, methods, subclasses).

I think this is fine because delegation is an **Implementation detail**. No user of the class should need that information unless the programmer decides it is relevant enough to add to the documentation.

> it can end up in the bottom of the class declaration. All of those scenarios dramatically harm the readability of the code.

Again, delegation is an **Implementation detail**. It is not critical enough that code is completely unreadable unless it is at the top of the class. The programmer themselves can decide whether a particular use of the feature is critical to the readability of the class and where to put it. In the case of `delegate val` it will be in the group of properties at the top of the class when adhering to the official Kotlin Coding Conventions.

> ```kotlin
> class X : A, B, C by C(), D by newD()
>    where A delegates a, B delegates this.b { ... // code left out }
> ```
> This approach simultaneously: simplifies the main class signature + keeps the delegation constructs at > the top + incorporates the old syntax as a short cut for some of the delegations + allows for a new > > semantics for the delegations that need recalculating their delegate. IMHO it's nearly perfect.

1. *simplifies the main class signature*
I think the class "signature" laid out here looks extremely convoluted, I would certainly not agree that it is "simplified".
Delegation *does not belong in the class signature* because it is an **implementation detail**.
2. *keeps the delegation constructs at the top*
See above.
3. *incorporates the old syntax as a short cut for some of the delegations*
The old syntax has many limitations/drawbacks in comparison to the new one. There should be no reason to use it when a new syntax is implemented. Therefore, it should probably be deprecated and we should at the very least discourage its usage. The old behaviour uses invisible fields and does not provide access to the Delegate Identity from source, among other problems, as explained in the file.
4. *allows for a new semantics for the delegations that need recalculating their delegate*
Well, obviously, if it didn't it wouldn't be a valid approach. I guess we agree here.

I think we should aim to:
* Avoid convoluting the class "signature"
* Treat delegation as an implementation detail, because that's what it is
* Replace the old behaviour - deprecating it - rather than actively supporting 2 features that aim to fill the same task

The last one of these points is the reason I created this proposal. That is what the proposal is for. To provide an alternative that is aimed at filling the same task, in a less limited, better, way. Once it's in place, the old behaviour is obsolete. It can be supported, but people should be encouraged to use the new option.

I might review the current text to make sure that that message is more clear.
Like @voddan pointed out earlier, it's not your job to convince me why a particular proposal might be better, I've accepted that and changed my mind / attitude. However, I would like to keep the goals of the proposals aligned.

**Comment by tieskedh** [comment]
I don't see it as new.
I see it as an extension of the current syntax.
That is why I still like to use `by`

**Comment by voddan** [comment]
@Dico200 
>  I would like to keep the goals of the proposals aligned.

Totally agree on that.

>  To provide an alternative that is aimed at filling the same task, in a less limited, better, way. Once it's in place, the old behaviour is obsolete

Not necessarily. If the new syntax **extends** the old syntax organically, then there is no need to deprecate the old syntax. It's a win-win.

> we could ask JB to give their opinion and point us in the right direction.

Yes, we need the help from JB in order to move forward (long before the KEEP is finalized). 

Also don't forget that this KEEP still needs a [shepherd](https://github.com/Kotlin/KEEP/#keep-shepherds). I think you, the KEEP's author, need to get one ASAP by writing to JB guys and/or advertising this KEEP on Slack in #language-propolsals.

**Comment by Dico200** [comment]
I updated the file to add 2 of the discussed syntaxes - `where` clause and `delegate <type> val` syntax.
Also, I made some small improvements. Will drop WIP marker because I think the file is up to date with the discussion now. Time to look for that shepherd.

**Comment by voddan** [comment]
> Will drop WIP marker because I think the file is up to date with the discussion now. Time to look for that shepherd.

👌

Should we post a link to the KEEP to the Slack channel? Then we can all "+1" it ;)

**Comment by Dico200** [comment]
@voddan You post it, I've posted the link several times before. I'll thumbs up your message!

@NKb03 
> `delegate A, B to getAB()` 

This syntax was suggested and is present in an earlier version of the file, I must have removed it because I thought `delegate val` syntax is significantly better. Although, `delegate val` syntax can be confusing in terms of the interface it's delegating. Maybe the `delegate ... to ...` syntax is a good approach after all. 
Adding property delegates into the mix does make sense, so you can use `by` instead of `to`, that's a cool idea. It can be confusing though because the two features are using a similar name and currently use the same keyword to declare.

> To reduce boilerplate there is also a `delegate` modifier usable on vals and vars which desugars ...

This is also a cool idea. Ability to use `delegate` keyword in multiple ways. It makes it more complex though and isn't strictly necessary, because `delegate T to a` can be expressed like `delegate val a: T`.
With the add benefit that you can never declare a delegate whose identity you can't access. With `delegate T to <expr>`, you don't really have a means of accessing `<expr>` from the source code without repeating it.

 > * Possible to keep delegate identities anonymous

You should only be able to do that by declaring visibility, which you can do with `delegate val`.

I'm not very convinced by the pros of `delegate T to a` over `delegate val` but at the same time, I don't think it's bad at all.

Will add it to the file again. Thanks for your feedback.

**Comment by NKb03** [comment]
The problem with `delegate val` is that you can only specify one type for the property.
So you three Interfaces:
`
interface A
interface B
interface C
`
And a class that implements them all:
`class X: A, B, C`
And you want to delegate only `A` and `B` to an instance of `X`

**Comment by Dico200** [comment]
@NKb03 The proposal lays out a few policies for this. Depending on the policy, you are correct.
> Possible type policies of a `delegate` property include:
> 1. the type is restricted to the interface types implemented by the class (mimicking old behaviour's effective policy)
> 1. the type is restricted such that there is always 1 common interface implemented by the type and the declaring class
> 1. the type is not restricted, and all its public members, except for those declared in `Any`, are delegated
> 1. the syntax can declare one (or optionally multiple) delegated interface type immediately following `delegate `,  
>    Then the property type is only restricted to implement all those types, like [2], but more explicit.

In the end, what you say applies to all proposals and the current feature. Delegation must be declared for each interface individually. The only exception can be `delegate val` at the moment, depending on the chosen policy. However, with `delegate val`, if a restrictive policy is used, it is arguably the least elegant indeed.

**Comment by tieskedh** [comment]
Any news?

**Comment by elizarov** [comment]
Thanks a lot for the discussion on this issue. However, it does not represent any specific proposal, so it is better suited for a discussion in the Kotlin issue tracker. I've updated the corresponding issue https://youtrack.jetbrains.com/issue/KT-83 with some basic requirements (but it does not have a specific design either). Please, cast your votes and continue the discussion there. 👍
